### ISSUE TITLE
BUG: Rounding floats which are already equal to an integer changes the value

### ISSUE THREAD

**Issue by FudgeMunkey** [numpy/numpy#20514]
### Describe the issue:

Running `np.round` adds `0.5` to some whole integer floats. For example, `3061040371728385.0` becomes `3061040371728385.0`. This error occurs if decimal values of `2, 5, 8` are used.

This becomes more significant with larger numbers such as `6.2768919806476296e16` which rounds up `8.0` numbers. This error occurs if decimal values of `1, 4, 7, 10` are used. Note, this isn't a problem if `62768919806476296` is used instead regardless of decimal value. 

*I checked decimcal values in the range `[0, 15]`

*This bug was found using [Hypothesis](https://hypothesis.readthedocs.io/).

### Reproduce the code example:

```python
A = np.array([3061040371728385.0])
out = np.round(A, 2)
diff = out - A # 0.5

assert_array_equal(out, A)

A = np.array([6.2768919806476296e16])
out = np.round(A, 1)
diff = out - A # 8.0

assert_array_equal(out, A)
```

### Error message:

```shell
# For np.array([3061040371728385.0]), 2
Mismatched elements: 1 / 1 (100%)
Max absolute difference: 0.5
Max relative difference: 1.63343158e-16
 x: array([3.06104e+15])
 y: array([3.06104e+15])

# For np.array([6.2768919806476296e16]), 1
Mismatched elements: 1 / 1 (100%)
Max absolute difference: 8.
Max relative difference: 1.27451612e-16
 x: array([6.276892e+16])
 y: array([6.276892e+16])
```

### NumPy/Python version information:

```
1.21.4 3.8.10 (default, Sep 28 2021, 16:10:42) 
[GCC 9.3.0]
```

**Comment by FudgeMunkey** [comment]
@Zac-HD

**Comment by matthew-brett** [comment]
Presumably due to:

```c
#include <stdio.h>
#include <math.h>

int main(int argc, char **argv) {
    double v = 3061040371728385.0;
    double rv = round(v * 100) / 100;
    printf("round 2dp of %f is  %f\n", v, rv);
}
```

Output:

```
round 2dp of 3061040371728385.000000 is  3061040371728385.500000
```

**Comment by seberg** [comment]
Yes, it can happen.  Should be the same issue as gh-13699.  Not sure if there is a good way to get around some of these paths at least.  Python seems to do slightly better, so maybe there is something to steal, if it doesn't make things much slower.

**Comment by Dharmatva** [comment]
I'd love to have a try at this issue, would that be fine with you guys? @seberg @matthew-brett @FudgeMunkey

**Comment by seberg** [comment]
Nobody will be working on this specifically. An improval proposal would be great!

**Comment by Dharmatva** [comment]
@matthew-brett, @seberg as you correctly identified, the error @FudgeMunkey stumbled upon can be attributed to limitations of the floating point representation (especially at larger scales). To get around it, we can treat the integral and the decimal portion of the input array separately and combine them *after* rounding the decimal portion. This leverages the fact that rounding works just fine at smaller scales.

This will ensure that:

1. Large errors aren't introduced when rounding numbers, especially at larger scale.
2. Rounding still works properly for smaller-scale floating point numbers

However, it is important to keep in mind that:

1. Errors cannot be entirely removed - only minimized using the following technique. 
2. Performance will take a hit as the following technique simply uses more operations than the existing version of `np.round`.

Here's my suggestion: implement two versions of round (as shown in the `updated_round` function below) - users may pass a specific parameter if they want a more 'precise' answer, otherwise numpy will default to the existing version of `np.round`

```python
import numpy as np

def updated_round(input_array, input_decimals, carefully_handle_large_numbers=False):
    
    if (carefully_handle_large_numbers):
        
        # separate the integral, decimal portion of the input
        decimal_portion = (input_array % 1)
        integral_portion = input_array - decimal_portion

        # round decimal portion *carefully*
        # (relies on the EXISTING version of np.round - which should be fine for small numbers)
        rounded_decimal_portion = np.round(decimal_portion  * (10 ** input_decimals)) / (10.0 ** input_decimals)

        # combine integral, decimal portion after rounding
        output_array = integral_portion + rounded_decimal_portion 
    
    else:
    
        # resort to normal rounding if user hasn't requested careful
        # handling of large numbers
        output_array = np.round(input_array, input_decimals)    

    return output_array
        
# EXAMPLE 1

A = np.array([3061040371728385.0])
old_out = updated_round(A, 2, carefully_handle_large_numbers=False)
new_out = updated_round(A, 2, carefully_handle_large_numbers=True)

old_diff = old_out - A # 0.5
new_diff = new_out - A # 0.0

print("EXAMPLE 1: ")
print(old_diff, new_diff)
print()

# EXAMPLE 2 

A = np.array([6.2768919806476296e16])
old_out = updated_round(A, 1, carefully_handle_large_numbers=False)
new_out = updated_round(A, 1, carefully_handle_large_numbers=True)

old_diff = old_out - A # 8.0
new_diff = new_out - A # 0.0

print("EXAMPLE 2: ")
print(old_diff, new_diff)
print()

# EXAMPLE 3 (just to demonstrate that rounding still works fine on small numbers)

A = np.array([6.6789])

old_out = updated_round(A, 3, carefully_handle_large_numbers=False)
new_out = updated_round(A, 3, carefully_handle_large_numbers=True)

old_diff = old_out - A # 0.0001
new_diff = new_out - A # 0.0001

print("EXAMPLE 3: ")
print(old_diff, new_diff)
print()
```

The above code has the following output:

```
EXAMPLE 1: 
[0.5] [0.]

EXAMPLE 2: 
[8.] [0.]

EXAMPLE 3: 
[0.0001] [0.0001]

```

**Comment by Dharmatva** [comment]
@seberg @matthew-brett @Zac-HD Hey guys, what do you think about my proposal?

**Comment by mattip** [comment]
In what proportion of the cases is there a difference in the output between the methods?

I think this would be relevant if we add a `dtype` kwarg to `np.round`, since currently the output is always float, and the inherent limits of float representation of integers will limit the usefulness of the change. As [mentioned in the other issue](https://github.com/numpy/numpy/issues/13699#issuecomment-948095503) I would be curious to hear what the use case is for rounding large floats.

**Comment by Dharmatva** [comment]
@mattip  I don't think such a scenario is too common 'in the wild' (but so is the case for many other bugs). However, here are some examples where it _might_ come into play:

- Someone trying to create discrete bins (for let's say, a histogram) over a range of values that are really large.
- Sensitive ML training algorithms relying on large loss values to iterate might use calculations that indirectly rely on `np.round`...
- Someone who deliberately just wants to try it, I guess?

But in any case, I think rounding shouldn't lead to massive errors in any case. Even if someone uses this function with a large value the code should at-least try to minimize errors. **If a change in code isn't warranted, there should at-least be a warning in documentation for `np.round` that shares some examples related to this anomalous behavior.**

**Comment by seberg** [comment]
A factor of two slower seems quite a lot, OTOH, I would not be surprised if we could have much faster rounding for certain cases (e.g. all cases without digits after the comma)?

**Comment by miccoli** [comment]
The problem here is how *round to `n` decimal places* is defined for floating point numbers with a non decimal base.

If I got it right, (but I still have to check the source), numpy uses a variation of the [round to specified multiple](https://en.wikipedia.org/wiki/Rounding#Rounding_to_a_specified_multiple) formula

$$
\mathrm {roundToMultiple} (x,m)=\mathrm {round} (x/m)\times m
$$

namely

$$
\mathrm {roundToDecimalPlaces} (x, n) = \mathrm {round} (x \times 10^n ) /  10^n
$$

or

$$
\mathrm {roundToDecimalPlaces} (x, n) = \mathrm {round} (x / 10^{-n} ) \times 10^{-n}
$$

The problem is that both the division $\cdot/ 10^n$ and the exponentiation $10^{-n}$ are inexact for binary floating numbers. Strictly speaking a more refined approach should be used, because the adopted approach can break in edge cases.

A related, much more serious, consequence of the way in which `np.round(a, decimals=d)` is defined for `d>0` is that numpy floats break the Liskov substitution principle.
```
>>> py_f = 2.765
>>> np_f = np.float_(py_f)
>>> assert np_f == py_f
>>> assert isinstance(np_f, float)
>>> assert round(np_f, 2) == round(py_f, 2)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AssertionError
```
In fact (remeber that `round(number)` delegates to `number__round__`)
```
>>> round(py_f, 2)
2.77
>>> round(np_f, 2)
2.76
```

Who is right? One could argue in favor of numpy (round to even), or python (2.77 is nearer to 2.765 than 2.76):
```
>>> round(py_f, 2) - py_f
0.004999999999999893
>>> round(np_f, 2) - np_f
-0.0050000000000003375
```

There are other problems with `np.round`, eg.  #11881, so maybe a general revamping of the code could be undertaken.
